cg/server/endpoints/error_handler.py

```diff
@@ -0,0 +1,27 @@
+import logging
+from functools import wraps
+from http import HTTPStatus
+
+from flask import jsonify
+
+from cg.store.exc import EntryNotFoundError
+
+LOG = logging.getLogger(__name__)
+
+
+def handle_endpoint_errors(func):
```

Comment:
This naming feels very general, whereas it is written to handle errors in a specific endpoint I would rename it to reflect that

Reply:
I made it to be used in any endpoint

Reply:
But that assumes the same error raising in all endpoints it is used on right?

Reply:
I think this makes sense for end points

Reply:
I do not see it being an endpoint as being a strong enough argument. While converting errors into a corresponding HTTP status code and error message is only wished for in endpoints it is not sufficient reasoning given that one endpoint could need completely different error handling despite the same error raised.

That being said, I think you could convince me to go either way with this. But if you go with having a vague naming and intending it to be reused for all endpoints, we should beware its limitations. Especially I think we should be ready to move away from the pattern if we run into a situation where the same exception needs to be handled in different ways for different endpoints. Writing this down, I am struggling to find good reasoning for aiming for a general error handler - what are the benefits?

Reply:
I will rename to `handle_missing_entries`

Reply:
I discussed a bit with chatgpt and seems as if the most common error handling could be in a common decorator and you could have separate endpoint specific ones if you need anything more special. So I think there are more than two options here.

My only point is that it is not clear to me that we will always want the same HTTP status returned for the same exception being caught. In this case it might make sense to have a 404 not found for the NoEntryFound error (or a 403 bad request or something else completely) but in another scenario the NoEntryFound might signal something different, or be something expected. 

Unless we are absolutely certain that all caught NoEntryFound errors should result in a 404 error code (which I am not), it just seems more adaptable for the future to not have it in a super general handler.

Reply:
> My only point is that it is not clear to me that we will always want the same HTTP status returned for the same exception being caught. In this case it might make sense to have a 404 not found for the NoEntryFound error (or a 403 bad request or something else completely) but in another scenario the NoEntryFound might signal something different, or be something expected.

My argument is, the logic or decisions to raise this error or what this error means should not be the responsibility of the end point.

I think the responsibility of the end point should be:
get a request
send a response
The response can be either data or the error message.

If for some reason a NoEntryFound error is expected given the request, then the layer below the endpoint should deal with this, not the end point.

Reply:
And where would you want the mapping between the exceptions and the corresponding status code to take place? Because this seems like a rebuttal of the need for this decorator altogether to me and that all this logic should take place in the service but then we are tying the service to only being used for a web service

Reply:
The mapping would be in the decorator.

The logic to do something different in instances where an EntryNotFound error should be re-interpreted  in the service

I actually like the idea of the ApplicationService becoming an ApplicationWebService - it reduces its scope and responsibility to only that, which I think is a good thing.

Reply:
I will resolve this and we can continue the discussion about error handlers somewhere else

Reply:
I agree much more with you than I think my comments convey - in general it is probably bad design if each endpoint has a different mapping for the same exception, say if the error type is too broad. And I am completely fine with this being a general error handler for the time being. I just think that there's legitimacy to questioning if we can use the same decorator for all methods in our API. Couple that with my failure to see any drawbacks with the alternative - to introduce endpoint specific handling in instances where it is suitable - I am just worried that we are looking at this from the wrong perspective. In my mind it is very much a situation where there are arguments for both sides, and the most problematic outcome would be if start writing code where good alternatives have been branded as something to be avoided. Definitely a discussion out of scope for this PR though.


Regarding doing a more narrow service, I could agree or disagree with you depending on what you want to include in it. The narrow scope I am positive of, but I would not want to see HTTP code conversions in it, since it would create an artificial hindrance for its functionality to be useful in other parts of the code.